### PR TITLE
Pulling up state, plus some refactoring

### PR DESCRIPTION
**Pull request by kekonen**
#622

### CODE REVIEW

File: src/client/components/CampaignContactEditor/CampaignContactList.react.js [hunk 1]
```diff
@@ -292,8 +278,8 @@ class CampaignContactList extends Components.ConditionalRenderComponent
                         className="-striped -highlight ocbn table"
                     />
                       <CampaignContactImporter
-                        campaignId={this.state.campaignId}
-                        customerId={this.state.customerId}
+                        campaignId={campaign.id}
+                        customerId={customer.id}
                         ref={node => this.importContent = node}
                         onResult={(err, res) => !err && this.reload()}
```
Comment (nczeranka) [inline src/client/components/CampaignContactEditor/CampaignContactList.react.js]: > this.reload()
does not exist anymore.
How to you update the contact list after an import? Guess you need to call a callback to set the contacts in upper component.

File: src/client/components/CampaignContactEditor/CampaignContactList.react.js [hunk 2]
```diff
@@ -14,16 +14,16 @@ import translations from './i18n';
...
+        customer : PropTypes.object.isRequired,
+        contacts : PropTypes.array.isRequired,
         disabled: PropTypes.bool,
+        updateCampaignState: PropTypes.func.isRequired,
     };
 
-    static defaultProps = {
-        isAnonymousCampaign : false,
-    };
+    // static defaultProps = {
```
Comment (nczeranka) [inline src/client/components/CampaignContactEditor/CampaignContactList.react.js]: Please remove code that you are not planning to user anymore.

File: src/client/components/CampaignContactEditor/CampaignContactList.react.js [hunk 3]
```diff
@@ -247,7 +233,7 @@ class CampaignContactList extends Components.ConditionalRenderComponent
                     <div className="col-sm-12 buttons">
                         <button className="btn btn-default pull-left"
                             onClick={e => this.handleEditContact(e)}
-                            disabled={this.state.isAnonymousCampaign || this.props.disabled}
+                            disabled={campaign ? campaign.invitationCode? true : false : false || this.props.disabled}
```
Comment (nczeranka) [inline src/client/components/CampaignContactEditor/CampaignContactList.react.js]: Interesting construct. Please simplify!

File: src/client/components/CampaignContactEditor/CampaignContactList.react.js [hunk 4]
```diff
@@ -256,13 +242,13 @@ class CampaignContactList extends Components.ConditionalRenderComponent
...
                         </button>
-                        <button className={"btn btn-default pull-right" + (contacts.length > 0 ? '' : 'disabled')} onClick={e => this.handleExport(e)}>
+                        <button className={"btn btn-default pull-right" + (contacts.length > 0 ? '' : ' disabled')} onClick={e => this.handleExport(e)}>
                             <span className="glyphicon glyphicon-export" />&nbsp;
                             {i18n.getMessage('CampaignContactList.button.export')}
                         </button>
                         <button className="btn btn-default pull-right"
                             onClick={e => this.handleUpload(e)}
-                            disabled={this.props.disabled}
+                            disabled={campaign ? campaign.invitationCode? true : false : false || this.props.disabled}
```
Comment (nczeranka) [inline src/client/components/CampaignContactEditor/CampaignContactList.react.js]: see comment above.

File: src/client/components/CampaignEditor/CampaignEditForm.react.js [hunk 5]
```diff
@@ -60,11 +61,12 @@ class CampaignEditForm extends Components.ConditionalRenderComponent
         super(props)
 
         const basicState = {
+            // campaign: props.campaign,
             customerId: props.customerId,
             errors: { }
         }
 
-        this.state = extend(false, { }, CampaignEditForm.emptyFormItem, basicState);
+        this.state = Object.assign({ }, CampaignEditForm.emptyFormItem, props.campaign || {}, basicState);
```
Comment (nczeranka) [inline src/client/components/CampaignEditor/CampaignEditForm.react.js]: Ok to replace extend - if it is working. But then please remove the extend include, too.

File: src/client/components/CampaignEditor/CampaignEditForm.react.js [hunk 6]
```diff
@@ -75,27 +77,15 @@ class CampaignEditForm extends Components.ConditionalRenderComponent
...
+    componentDidUpdate(prevProps, prevState)
     {
-        if(nextPops.campaignId != this.props.campaignId)
-            return this.loadCampaign(nextPops.campaignId);
-    }
-
-    reload()
-    {
-        return this.loadCampaign(this.props.campaignId);
+        if (JSON.stringify(prevProps.campaign) !== JSON.stringify(this.props.campaign)) {
```
Comment (nczeranka) [inline src/client/components/CampaignEditor/CampaignEditForm.react.js]: Not sure about the need for this check (which is btw. not recommended by react). Isn't it enough to check the id?

File: src/client/components/CampaignEditor/CampaignEditForm.react.js [hunk 7]
```diff
@@ -305,8 +287,9 @@ class CampaignEditForm extends Components.ConditionalRenderComponent
                 })
                 .then(() => this.campaignsApi.updateCampaign(campaign.id, campaign))
                 .then(item => {
-                    this.putItemToState(item);
-                    this.formChanged = false;
+                    this.props.updateCampaignState({campaign: item})
+                    // this.putItemToState(item);
```
Comment (nczeranka) [inline src/client/components/CampaignEditor/CampaignEditForm.react.js]: Please remove, if not used anymore.

File: src/client/components/CampaignEditor/CampaignEditForm.react.js [hunk 8]
```diff
@@ -322,8 +305,9 @@ class CampaignEditForm extends Components.ConditionalRenderComponent
                     else {
                         return this.campaignsApi.addCampaign(campaign)
                         .then(item => {
-                            this.putItemToState(item);
-                            this.formChanged = false;
+                            this.props.updateCampaignState({campaign: item})
+                            // this.putItemToState(item);
```
Comment (nczeranka) [inline src/client/components/CampaignEditor/CampaignEditForm.react.js]: Please remove, if not used anymore.

File: src/client/components/CampaignEditor/CampaignEditForm.react.js [hunk 9]
```diff
@@ -336,12 +320,12 @@ class CampaignEditForm extends Components.ConditionalRenderComponent
         }
     }
 
-    clearForm()
-    {
-        this.putItemToState(CampaignEditForm.emptyFormItem);
-        this.setState({ errors : { } });
-        this.formChanged = false;
-    }
+    // clearForm()
```
Comment (nczeranka) [inline src/client/components/CampaignEditor/CampaignEditForm.react.js]: Please remove if not used anymore.

File: src/client/containers/Campaign.react.js [hunk 10]
```diff
@@ -22,19 +22,20 @@ class Campaign extends Components.ConditionalRenderComponent
         super(props);
 
         this.state = {
+            customer: context.userData.customerid ? {id: context.userData.customerid}: {},
+            campaign: context.router.params.campaignId ? {id: parseInt(context.router.params.campaignId)} : {},
+            contacts: [],
+
```
Comment (nczeranka) [inline src/client/containers/Campaign.react.js]: To be complete would be helpfull to include the two template class variables, too.

File: src/client/containers/Campaign.react.js [hunk 11]
```diff
@@ -56,87 +57,57 @@ class Campaign extends Components.ConditionalRenderComponent
...
+        context.i18n.register('Campaign', translations);
+        context.i18n.register('CampaignContact', translationsService);
+        context.i18n.register('CampaignContact', translationsCampaignContact);
 
-        return this.init(this.context.router.params.campaignId);
+        this.updateCampaignState = this.updateCampaignState.bind(this);
     }
 
-    componentWillReceiveProps(nextPops, nextContext)
+    updateCampaignState(state, extra)
```
Comment (nczeranka) [inline src/client/containers/Campaign.react.js]: It is not easy to read and to understand all the ways you are using it.
Please simplify by creating explicit methods for the different use cases.

File: src/client/containers/CampaignTemplateSelection.react.js [hunk 12]
```diff
@@ -4,26 +4,26 @@ import { Components } from '@opuscapita/service-base-ui';
 import { Api, TemplatePreview, TemplateForm, FileManager } from '../components/TemplateEditor';
 import { Campaigns } from '../api';
 import translations from './i18n';
-import extend from 'extend';
+// import extend from 'extend';
```
Comment (nczeranka) [inline src/client/containers/CampaignTemplateSelection.react.js]: Please remove if not used anymore.

File: src/client/containers/CampaignTemplateSelection.react.js [hunk 13]
```diff
@@ -4,26 +4,26 @@ import { Components } from '@opuscapita/service-base-ui';
...
+        templates : PropTypes.array,
+        updateCampaignState: PropTypes.func.isRequired,
     }
 
     static defaultState = {
         type : '',
-        campaignId : '',
         campaign : { },
-        customerId : '',
+        customerId : "",
```
Comment (nczeranka) [inline src/client/containers/CampaignTemplateSelection.react.js]: Still used? Shouldn't it be replaced by the passed property?

File: src/client/containers/CampaignTemplateSelection.react.js [hunk 14]
```diff
@@ -4,26 +4,26 @@ import { Components } from '@opuscapita/service-base-ui';
...
         templateFileDirectory : PropTypes.string,
-        selectedTemplate : PropTypes.string
+        templates : PropTypes.array,
+        updateCampaignState: PropTypes.func.isRequired,
     }
 
     static defaultState = {
         type : '',
-        campaignId : '',
         campaign : { },
```
Comment (nczeranka) [inline src/client/containers/CampaignTemplateSelection.react.js]: Still used?
Please check also for usages - see e.g. line 92!

File: src/client/containers/CampaignTemplateSelection.react.js [hunk 15]
```diff
@@ -41,38 +41,49 @@ class CampaignTemplateSelection extends Components.ConditionalRenderComponent
         this.validationMessage = null;
         this.previews = [ ];
 
-        const basicState = {
-            selectedTemplate : this.props.selectedTemplate,
-            templates : [ ]
-        };
+        // const basicState = {
```
Comment (nczeranka) [inline src/client/containers/CampaignTemplateSelection.react.js]: Please remove code that you are not using anymore.

File: src/client/containers/CampaignTemplateSelection.react.js [hunk 16]
```diff
@@ -244,7 +230,8 @@ class CampaignTemplateSelection extends Components.ConditionalRenderComponent
                 const successMessage = i18n.getMessage('CampaignTemplateSelection.notification.template.delete.success');
 
                 return Api.deleteTemplate(this.props.customerId, templateId, this.templateFileDirectory)
-                    .then(() => this.reload())
+                    // .then(() => this.reload())
```
Comment (nczeranka) [inline src/client/containers/CampaignTemplateSelection.react.js]: Please remove code that you are not using anymore.

Comment (kekonen) [reply src/client/containers/CampaignTemplateSelection.react.js]: done

File: src/client/containers/CampaignTemplateSelection.react.js (hunk 15 above)
Comment (kekonen) [reply src/client/containers/CampaignTemplateSelection.react.js]: done

File: src/client/containers/CampaignTemplateSelection.react.js (hunk 14 above)
Comment (kekonen) [reply src/client/containers/CampaignTemplateSelection.react.js]: OK, done

File: src/client/containers/CampaignTemplateSelection.react.js (hunk 13 above)
Comment (kekonen) [reply src/client/containers/CampaignTemplateSelection.react.js]: done

**Pull request merged**